Vim mode

sijieamoy opened the pull request:
## TLDR
Implements a subset of vim keybindings under `/vim` mode. 

## Dive Deeper

### **Core Vim Functionality**
- **Dual mode system**: NORMAL and INSERT modes with proper transitions
- **Navigation commands**: `h`, `j`, `k`, `l` (directional), `w`, `b`, `e` (word movement), `0`, `$`, `^` (line navigation), `G`, `gg` (line jumping)
- **Editing commands**: `x` (delete), `i`, `a`, `o`, `O` (insert modes), `I`, `A` (line insert modes), `d`, `c` (delete/change operators), `D`, `C` (line operations)
- **Complex operations**: Multi-key commands like `dd`, `cc`, `dw`, `cw`, `db`, `cb`, `de`, `ce`
- **Count support**: All commands support numeric prefixes (e.g., `3h`, `5w`, `10G`)
- **Repeat functionality**: `.` command repeats the last editing operation

### **Settings & Persistence**
- **Persistent configuration**: Vim mode preference saved in `~/.gemini/settings.json`
- **Runtime toggling**: `/vim` slash command for on-the-fly mode switching
- **Status indicator**: Visual `[NORMAL]`/`[INSERT]` indicator in footer
- **Session restoration**: Vim mode state preserved between CLI restarts

### **Integration & Compatibility**
- **Tab completion**: Full support for file/command completion in INSERT mode
- **Arrow key navigation**: Up/down arrows navigate completion suggestions in both modes
- **Input delegation**: Clean architecture separating vim handling from completion logic
- **Escape handling**: Proper escape sequence detection that doesn't interfere with arrow keys
- **Image pasting**: Supports the same Ctrl+V to paste image feature

### Documentation
- **Commands reference**: Added `/vim` command to CLI commands documentation
- **Configuration guide**: Added `vimMode` setting to configuration documentation
- **Feature coverage**: Comprehensive documentation of all supported vim operations

### Testing
- **Unit tests**: 38 vim functionality tests 
- **Regression testing**: All existing CLI functionality verified
- **Manual testing**: Extensive testing of vim command sequences and edge cases



## Reviewer Test Plan
1. Launch Gemini CLI
2. Use `/vim` to turn on the Vim mode
3. Test the subset of Vim keybindings listed above.

## Linked issues / bugs
Fixes #1893

jacob314 commented on packages/cli/src/config/config.ts:
```diff
@@ -438,6 +438,7 @@ export async function loadCliConfig(
     listExtensions: argv.listExtensions || false,
     extensions: allExtensions,
     blockedMcpServers,
+    vimMode: settings.vimMode,
```
revert this as well as it isn't needed as this can just be read off settings directly.

jacob314 commented on packages/cli/src/ui/hooks/vim.ts:
```diff
@@ -0,0 +1,1571 @@
+
+              // Delete from cursor through N words using findNextWordStart for consistency
+              let searchOffset = currentOffset;
+              for (let i = 0; i < repeatCount; i++) {
```
this logic is not safe to perform here. instead use the existing api in text-buffer.ts or create one. that. is crucial to make sure this logic runs inside the reducer in text-buffer that ensures commands are executed properly and in order even when multiple key strokes are executed in the same react loop. the same concern applies to all of the other cases here that modify the state of the text. buffer. they need to instead call a method on text-buffer.ts to apply the transformation desired rather than reading the possibly stale state of the text buffer and performing operations on it.
sorry this part of how text-buffer works is a little surprising but it is better than the alternatives where it becomes really difficult to be confident that react is dealing with the correct state.

jacob314 reviewed (commented):
I'm done reviewing all aspects of this pull request that don't involve ensuring the Vim bindings themselves are what Vim users expect. I'll defer to @sethtroisi for that who uses Vim on a daily basis and should have a better idea about that part of this pull request.  Overall this pull request looks really good and I'm quite confident we will be able to land it after these comments are addressed. 

Please do push back on some of the refactorings suggested for vim.ts. Those refactorings would be ideal but if I'm missing some things that make them infeasible I'm open to leaning more heavily into useRef to achieve the effects desired rather than requiring use of reducers, etc.

sijieamoy commented on packages/cli/src/ui/App.tsx:
```diff
@@ -388,6 +389,7 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
     openAuthDialog,
     openEditorDialog,
     toggleCorgiMode,
+    () => toggleVimModeRef.current?.(), // Wrapper function that calls the ref
```
With the help of Gemini CLI, I considered two options:
1. Create a `VimContext` that wraps the App component
2. Pass a generic settings update function instead of the specific `toggleVimMode`.

Ended up going with (2) because the amount of changes is smaller, and the `VimContext` feels an overkill because the rest of the app will unlikely need to use the context. Let me know if this makes sense.

sijieamoy commented on packages/cli/src/ui/commands/vimCommand.ts:
```diff
@@ -0,0 +1,20 @@
+
+    toggleVimMode();
+
+    // No message - silent toggle
```
Added a message when enabling the Vim mode. I agree it's a good idea to show the basics of mode toggling just in case someone accidentally enables it. Didn't add the exit message though given the lower ROI.
<img width="614" height="78" alt="image" src="https://github.com/user-attachments/assets/67a2564a-63f6-461e-bb67-bca50e63b718" />

sijieamoy commented on packages/core/src/config/config.ts:
```diff
@@ -514,6 +517,10 @@ export class Config {
     return this.bugCommand;
   }
 
+  getVimMode(): boolean {
```
Done.

sijieamoy commented:
> I'm done reviewing all aspects of this pull request that don't involve ensuring the Vim bindings themselves are what Vim users expect. I'll defer to @sethtroisi for that who uses Vim on a daily basis and should have a better idea about that part of this pull request. Overall this pull request looks really good and I'm quite confident we will be able to land it after these comments are addressed.
> 
> Please do push back on some of the refactorings suggested for vim.ts. Those refactorings would be ideal but if I'm missing some things that make them infeasible I'm open to leaning more heavily into useRef to achieve the effects desired rather than requiring use of reducers, etc.

Appreciate the feedback! Applied most, except for the Vim mode exit message.

sijieamoy commented on packages/cli/src/ui/components/InputPrompt.test.tsx:
```diff
@@ -130,6 +130,7 @@ describe('InputPrompt', () => {
       config: {
         getProjectRoot: () => '/test/project',
         getTargetDir: () => '/test/project/src',
+        getVimMode: () => false,
```
Added three to make sure the right handler is called in INSERT/NORMAL/disabled modes.

jacob314 commented:
I'm seeing issues with the message that you are in vim mode not displaying until you press another character. Can you fix that?
https://screencast.googleplex.com/cast/NjAxMjk3MjEwNjI1MjI4OHxjOTUxNDE1OS00ZQ 

I'd suggest we do add a message to the history as well as changing the default text shown in the input prompt when entering and exiting vim mode. Otherwise it isn't really clear whether you turned vim mode on and off. I care a lot that users who typed `/vim` see a message explaining how to use it and how to turn it off if they didn't realize they just did something that is changing their `settings.json`.

jacob314 commented on packages/cli/src/ui/commands/types.ts:
```diff
@@ -49,6 +49,13 @@ export interface CommandContext {
     toggleCorgiMode: () => void;
+    /** Updates a setting in the application configuration. */
+    updateSetting: <
+      K extends keyof import('../../config/settings.js').Settings,
```
fix this type definition to not inline the imports here.

jacob314 commented on packages/cli/src/ui/hooks/vim.ts:
```diff
@@ -0,0 +1,1420 @@
+        }
+
+        case CMD_TYPES.DELETE_CHAR: {
+          for (let i = 0; i < count; i++) {
```
needs to be moved to textBuffer

jacob314 replied:
this comment and the others in vim.ts along the same lines need to be addressed to make this code robust. this is the root cause of issues like the one you were seeing where pressing escape followed by `w` not behaving correctly. Right now due to this issue you could see `esc` followed by two `w` only behaving like `esc` followed by one `w`.
You can also reproduce this sort of issue by using ssh on a slow terminal.

jacob314 commented on packages/cli/src/ui/hooks/vim.ts:
```diff
@@ -0,0 +1,1420 @@
+  /** Handles end-of-line operations (D, C) */
+  const handleEndOfLineOperation = useCallback(
+    (shouldEnterInsertMode: boolean) => {
+      const currentRow = buffer.cursor[0];
```
the conditional logic in this method needs to go into textBuffer as it will otherwise have the stale issues discussed above

jacob314 commented on packages/cli/src/ui/hooks/vim.ts:
```diff
@@ -0,0 +1,1420 @@
+        }
+
+        case CMD_TYPES.DELETE_LINE: {
+          const startRow = buffer.cursor[0];
```
this needs to go into textBuffer as a deleteLine helper

jacob314 reviewed (commented):
This is looking quite good. Remaining blockers are just cases where move of the logic manipulating the text-buffer needs to go into text-buffer.ts to play well with the reducer. Good news is all the cases still in vim.ts look like they should be fairly clean to move to text-buffer.ts and when they are moved there will also be nice if someone wants to add non-vim keyboard shortcuts for those operations in the future.

sijieamoy commented:
@jacob314 a few major changes:
1. RE: [your previous comment on avoiding toggleVimModeRef](https://github.com/google-gemini/gemini-cli/pull/3936/#discussion_r2223583925), I ended up switching to adding a VimMode context because the settings approach made mode change propagation very tricky, React isn't great at detecting settings changes either. Hope this makes sense.
2. Moved all the buffer manipulation code to text-buffer.
3. Added enter/exit messages to the message list.

@sethtroisi thanks for the review as well! The `e` and arrow key issues are both fixed.

jacob314 commented on packages/cli/src/ui/hooks/useKeypress.ts:
```diff
@@ -147,12 +147,15 @@ export function useKeypress(
-      rl = readline.createInterface({ input: keypressStream });
+      rl = readline.createInterface({
+        input: keypressStream,
+        escapeCodeTimeout: 0,
```
I  think this may have been an issue resolved by other changes you landed to switch to a reducer and better align the operations with text-buffer. I've locally set this timeout to 2000 and I can't reproduce the issue. before this code was using a reducer handling multiple commands in a single react event loop would cause confusing bugs which I suspect is what you are seeing. As I understand it, settingescapeCodeTimeout to 0 should only change this behavior by ensuring you did not see a escape and a regular key both fired within the same react event loop. Apologies if I am missing something and there is a real issue here. If not please revert the escapeCodeTimeout change and I will approve. 
Fyi @bbiggs who I believe also considered lowering the escapeCodeTimout as the default value is quite high. We might consider a sigificantly lower value than the default although I am not sure we shoudl go all the way to zero.

jacob314 commented on packages/cli/src/ui/hooks/vim.ts:
```diff
@@ -0,0 +1,863 @@
+  );
+
+  // Helper to get current cursor position as offset
+  const getCurrentOffset = useCallback(() => {
```
remove this helper completely as it is not safe given the fact that you want to handle multiple keystrokes one after the other so the buffer could be stale by the time you handle the second one if both keystrokes are handled in the same react event frame. sorry this logic has to be written in such a particular way to be robust but you are making changes to the trickiest part of the input system where we need to chain operations together carefully to avoid confusing user behavior.

sijieamoy replied:
Sorry, what do you mean by "remove"? Or, in other words, what's the proper way to get the current offset in vim.ts?

I wrote a simple AppleScript to simulate commands like `Dw` and `ddw` in quick succession (near zero delay), and it does reveal the problem of the 2nd command operating on stale text buffer. 

I experimented with `useRef(buffer)` in vim.ts, so that the input handler can get the fresh buffer data after the 1st command is executed in the same frame. It appears to have fixed the staleness for `Dw`, `esc+w`, `esc+3x` etc., but not for `ddw` - `w` would still be handled with stale cursor pos instead of pos 0 of the next line after `dd`. I guess it might be a race condition between the handling of `dd` (deletion and cursor pos calculation) and `w`. If there is a 10ms delay between `dd` and `w` it actually works fine.

sijieamoy replied:
Removed all the text buffer state reading and direct manipulation from vim.ts.

jacob314 reviewed (commented):
approved after these comments are addressed. if there are any more state issues after that I'll go ahead and take a look at polishing the reducer logic in vim.ts as I think about it more I think all the logic in vim.ts might need to be within the reducer rather than on the side of reducer otherwise we aren't really handling the multiple keystrokes within the same event loop cases robustly. to check that we should really provide some ways to make it easier for you to test that so please don't let that block you from landing this.![lgtm](https://i.imgur.com/JgNRalI.png)

jacob314 reviewed (approved):
![lgtm](https://i.imgur.com/JgNRalI.png)
Thanks for all the polish to make vim.ts more robust!

ali-aljufairi commented:
@sijieamoy wow such a cool contribution thanks for taking time to do it and also thanks for also making it being in setting.json I will merge main and try it out but thanks

ljxfstorm commented:
This PR's changes to `package-lock.json` will introduce missing `resolved` and `integrity` fields for some dependencies. This is known to cause issues like #4109.
Specifically, it's the deletions between lines 11707 to 11921 in `package-lock.json`.
Could you please restore the deleted sections?
